pkg/controller/noderef/noderef_controller.go

```diff
@@ -135,6 +135,10 @@ func (r *ReconcileNodeRef) Reconcile(request reconcile.Request) (reconcile.Resul
 
 	result, err := r.reconcile(ctx, cluster, machine)
 	if err != nil {
+		if err == ErrNodeNotFound {
+			return result, nil
```

Comment:
Hi @pires,

You [asked](https://github.com/kubernetes-sigs/cluster-api/pull/1117#issuecomment-508406970)

> I'm trying to understand how this code could affect the failing test.

The [failure](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/sigs.k8s.io_cluster-api/1117/pull-cluster-api-test/1146532328546242562/) is:

```shell
I0703 21:35:53.413] === RUN   TestReconcile
I0703 21:35:53.414] E0703 21:35:48.331304   15023 machinedeployment_controller.go:138] Failed to reconcile MachineDeployment "default/foo": Operation cannot be fulfilled on machinedeployments.cluster.sigs.k8s.io "foo": the object has been modified; please apply your changes to the latest version and try again
I0703 21:35:53.414] --- FAIL: TestReconcile (5.49s)
I0703 21:35:53.414]     machinedeployment_reconciler_test.go:150: timed out waiting reconcile error
I0703 21:35:53.415] FAIL
I0703 21:35:53.415] FAIL	sigs.k8s.io/cluster-api/pkg/controller/machinedeployment	15.051s
I0703 21:35:53.415] FAIL	sigs.k8s.io/cluster-api/pkg/controller/machinedeployment	15.051s
```

So it's timed out waiting. The one change that I see that may affect that is this line:

```golang
return result, nil
```

Previously if the node was not found, the same result would be returned (a Requeue), but also the error. Here the error is being swallowed and only the result is returned.

I do think the basic controller design is anti-pattern for Go since the language convention is to never return a non-empty value for return arguments when the error is non-nil. In the case of the controller design, the `reconcile.Result` is used to return a `RequeueAfter` `time.Duration` in addition to possible errors. It makes things difficult to follow IMO. I'm glad those that designed the `Actuator` interface saw fit to make the requeue semantic part of the actual `error`.

Anyway, I can't say for certain this is the line causing the problems, but this line **does** change the behavior of the system. Where once `reconcile.Result{RequeueAfter: 10 * time.Second}` was returned along with a non-nil `error`, now only the result is returned, and the `error` is omitted.

Reply:
There must be some misunderstanding here, because this PR fixes `reconcile()` so that `ErrNodeNotFound` is indeed passed on to the caller, `Reconcile()`.

Now, looking at [the controller-runtime call to Reconcile](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L216-L235), clearly, if we want the reconcile to be retriggered some time later, as is explicitly defined in the `reconcile.Result{RequeueAfter: 10 * time.Second}`, then `Reconcile()` must not return the error. If the error is returned in the `Reconcile()` call, the reconcile handler will not requeue.

Reply:
Hi @pires,

I do not disagree with any of what you said. I'm simply stating that in the `Reconcile` function, the only significant difference that I see is the error is no longer being returned.

Now, if that is causing requeues ad-infinitum, and that's the desired behavior, then perhaps you've exposed a bug as a side-effect of the change. Essentially the node is never found, the requeues keep checking, and so eventually the timeout occurs.

Reply:
Hi @pires,

I also think this may be an issue:

```shell
I0703 21:35:53.414] E0703 21:35:48.331304   15023 machinedeployment_controller.go:138] Failed to reconcile MachineDeployment "default/foo": Operation cannot be fulfilled on machinedeployments.cluster.sigs.k8s.io "foo": the object has been modified; please apply your changes to the latest version and try again
```

It may be that since the NodeRefController and MachineController (and variants) both modify a Machine's status, that something either is trying to do is failing because of a modification, and that that failure should be retried.

Again, I'm not sure your changes are the cause, but maybe have exposed an existing bug in the test design/controller. I can't say for sure -- Vince should take a look. He'll be back online tomorrow I think.

Reply:
I'm sorry but I still disagree with
> the only significant difference that I see is the error is no longer being returned.

Unless I'm failing miserably to understand my own changes, `ErrNotFound` was not returned before from `reconcile()` to caller `Reconcile()`, and therefore the latter would never return the same error to its caller.

Reply:
Hi @pires,

I apologize, perhaps I'm missing something as well. Let me take another look :)

Reply:
>  I'm not sure your changes are the cause, but maybe have exposed an existing bug in the test design/controller.

I am inclined to think the same and hopefully someone will be able to explain what's going on.

Reply:
Hi @pires,

Okay, you're absolutely correct. Value wise, nothing has changed with respect to the actual return values. This is nuts. As we discussed on Slack, it seems probable that some change to `master` since your PR was created is now causing a bug with this test. I'm going to attempt to run this test locally with your branch instead of allowing the CI to merge this with `master` and do the test there. If that works, then at least we know the reason for the failure, and at that point can start, or help someone start, track down the underlying cause.

Reply:
Hi @pires,

There's no question -- it's something already merged into `master` causing the issue. I checked our your branch and ran the following:

```shell
$ git log --oneline -1 && printf '\n\n' && { { ./scripts/ci-test.sh && printf '\nSUCCESS!\n'; } || printf '\nFAILED!\n'; }
```

It succeeded without error ([log](https://gist.github.com/akutz/c7a413bde11789cd1b9ad77cf9fc8ebb)).

I'm going to assign this to Vince to take a look when he's back.

/assign @vincepri

Reply:
Okay, this is weird. This branch is already rebased onto the current tip of `master`. This is just weird at this point.

Reply:
Hi @pires,

Okay, the test passed. It's clear now that it's a race condition where the `NodeRef` and `MachineDeploymentController` were both modifying a machine, and the second modification from the `MachineDeploymentController` failed, causing something to cause the reconciliation to fail since the node/machine relationship was never discovered.

tl;dr The test appears to be flaky

Reply:
FWIW, the following log prints an error, but is unrelated to the original failure:

```shell
machinedeployment_reconciler_test.go:204: error updating machineset: Operation cannot be fulfilled on machinesets.cluster.sigs.k8s.io "foo-6585bdd8bd": the object has been modified; please apply your changes to the latest version and try again
```

This message occasionally is seen when testing the `MachineDeploymentController`. To force this, run the tests in a loop:

```shell
$ for n in {1..20}; do printf 'BEGIN TEST RUN %d\n\n' "${n}" && go test -tags=integration -v -count 1 ./pkg/controller/machinedeployment; printf '\nEND TEST RUN %d\n\n' "${n}"; done
```

I've managed to cause the message to occur once or twice in a loop of twenty tests.

Please note that `go test -count 20` cannot be used because it re-uses the same process, and errors about objects already existing will occur.

Reply:
Additionally, here's [a gist](https://gist.github.com/akutz/d7191250a55ffa1617ad37e5ca7bac76) with instructions (and Dockerfile) for how to show that low-resource availability leads to flaky tests. I think that's what happened in @pires original test failure. Perhaps increasing the resources available for the Prow job definitions can help avoid this.

Reply:
My apologies for just now seeing this, this test is indeed flaky and has been for a while. 

@akutz thank you for tracking down the cause of this flake.

Reply:
@vincepri has made a change to test-infra to increase the number of cpus requested for some of our CI jobs. Hopefully this will fix the flakes.

Reply:
Rebased master which seems to have fixed this.